File: average-of-levels-in-binary-tree/average-of-levels-in-binary-tree.py
from collections import deque
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfLevels(self, root: TreeNode) -> List[float]:
        if not root:
            return []
        
        q = deque([(root, 0)])
        output = []
        
        sumVal = 0
        numNodes = 0
        
        while q:
            node, level = q.popleft()
            
            if level > len(output):
                output.append(sumVal/numNodes)
                
                sumVal = 0
                numNodes = 0
            
            sumVal += node.val
            numNodes += 1
            
            for child in [node.left, node.right]:
                if child:
                    q.append((child, level+1))
        
        output.append(sumVal/numNodes)
        return output

File: average-of-levels-in-binary-tree/average-of-levels-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Double> averageOfLevels(TreeNode root) {
        List<Double> output = new ArrayList<>();
        if (root == null) {
            return output;
        }
        
        int total = 0;
        long sum = 0;
        
        Deque<TreeNode> q = new ArrayDeque<TreeNode>();
        Deque<Integer> levels = new ArrayDeque<>();
        q.addLast(root);
        levels.addLast(0);
          
        while (!q.isEmpty()) {
            TreeNode node = q.pollFirst();
            int level = levels.pollFirst();

            
            if (level > output.size()) {
                output.add(1.0 * sum / total);
                total = 0;
                sum = 0;
            }
            
            sum += node.val;
            total++;
            
            for (TreeNode child : new TreeNode[]{node.left, node.right}) {
                if (child != null) {
                    q.addLast(child);
                    levels.addLast(level + 1);
                }
            }
            
        }
        output.add(1.0 * sum / total);
            
        return output;
    }
}